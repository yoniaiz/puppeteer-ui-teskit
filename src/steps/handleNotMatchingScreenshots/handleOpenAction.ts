import {
  buildDiffImagePath,
  saveDiffImage,
} from '../../utils/testkit.utils.js';
import open from 'open';
import { createDiffImage } from './createDiffImage.js';
import { FailedSnapshot } from '../../types/index.js';

export async function handleOpenAction({
  diffImageBuffer,
  filePath,
  snapName,
  newSnapImageBuffer,
  oldSnapImageBuffer,
}: FailedSnapshot) {
  const diffsImage = await createDiffImage({
    diffImageBuffer,
    newSnapImageBuffer,
    oldSnapImageBuffer,
  });

  await saveDiffImage({
    filePath,
    snapName,
    buffer: diffsImage,
  });
  await open(buildDiffImagePath(filePath, snapName));
}
