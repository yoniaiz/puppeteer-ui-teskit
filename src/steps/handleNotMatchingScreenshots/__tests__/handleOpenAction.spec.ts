import sharp from 'sharp';
import open from 'open';
import { handleOpenAction } from '../handleOpenAction.js';
import { fsMocks } from '../../../test/fsMocks.js';
import { pathMocks } from '../../../test/pathMocks.js';

jest.mock('sharp', () => ({
  __esModule: true,
  default: jest.fn(() => ({
    composite: jest.fn(() => ({
      toBuffer: jest.fn(() => Buffer.from('')),
      png: jest.fn(() => ({
        toBuffer: jest.fn(() => Buffer.from('')),
      })),
    })),
  })),
}));

jest.mock('open', () => jest.fn());

describe('handleOpenAction', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should handle opening the diff image', async () => {
    const mockDiffImageBuffer = Buffer.from('diff image data');
    const mockNewSnapImageBuffer = Buffer.from('new snapshot image data');
    const mockOldSnapImageBuffer = Buffer.from('old snapshot image data');

    pathMocks.resolve();
    fsMocks.promises.writeFile();
    const mockGetBottomBarBuffer =
      fsMocks.promises.readFile(mockDiffImageBuffer);

    const mockFailedSnapshot = {
      diffImageBuffer: mockDiffImageBuffer,
      filePath: 'path/to/file',
      snapName: 'snapshot',
      newSnapImageBuffer: mockNewSnapImageBuffer,
      oldSnapImageBuffer: mockOldSnapImageBuffer,
    };

    await handleOpenAction(mockFailedSnapshot);

    expect(mockGetBottomBarBuffer).toHaveBeenCalledTimes(1);

    expect(sharp).toMatchSnapshot();

    expect(open).toHaveBeenCalledTimes(1);
    expect(open).toHaveBeenCalledWith('path/to/snapshot.diff.png');
  });
});
