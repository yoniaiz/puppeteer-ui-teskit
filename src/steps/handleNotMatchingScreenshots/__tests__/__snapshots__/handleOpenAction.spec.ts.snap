// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handleOpenAction should handle opening the diff image 1`] = `
[MockFunction] {
  "calls": [
    [
      {
        "create": {
          "background": {
            "b": 0,
            "g": 0,
            "r": 0,
          },
          "channels": 3,
          "height": 1124,
          "width": 3340,
        },
      },
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": {
        "composite": [MockFunction] {
          "calls": [
            [
              [
                {
                  "gravity": "northeast",
                  "input": {
                    "data": [
                      110,
                      101,
                      119,
                      32,
                      115,
                      110,
                      97,
                      112,
                      115,
                      104,
                      111,
                      116,
                      32,
                      105,
                      109,
                      97,
                      103,
                      101,
                      32,
                      100,
                      97,
                      116,
                      97,
                    ],
                    "type": "Buffer",
                  },
                },
                {
                  "gravity": "north",
                  "input": {
                    "data": [
                      100,
                      105,
                      102,
                      102,
                      32,
                      105,
                      109,
                      97,
                      103,
                      101,
                      32,
                      100,
                      97,
                      116,
                      97,
                    ],
                    "type": "Buffer",
                  },
                },
                {
                  "gravity": "northwest",
                  "input": {
                    "data": [
                      111,
                      108,
                      100,
                      32,
                      115,
                      110,
                      97,
                      112,
                      115,
                      104,
                      111,
                      116,
                      32,
                      105,
                      109,
                      97,
                      103,
                      101,
                      32,
                      100,
                      97,
                      116,
                      97,
                    ],
                    "type": "Buffer",
                  },
                },
                {
                  "gravity": "south",
                  "input": {
                    "data": [
                      100,
                      105,
                      102,
                      102,
                      32,
                      105,
                      109,
                      97,
                      103,
                      101,
                      32,
                      100,
                      97,
                      116,
                      97,
                    ],
                    "type": "Buffer",
                  },
                },
              ],
            ],
          ],
          "results": [
            {
              "type": "return",
              "value": {
                "png": [MockFunction] {
                  "calls": [
                    [],
                  ],
                  "results": [
                    {
                      "type": "return",
                      "value": {
                        "toBuffer": [MockFunction] {
                          "calls": [
                            [],
                          ],
                          "results": [
                            {
                              "type": "return",
                              "value": {
                                "data": [],
                                "type": "Buffer",
                              },
                            },
                          ],
                        },
                      },
                    },
                  ],
                },
                "toBuffer": [MockFunction],
              },
            },
          ],
        },
      },
    },
  ],
}
`;
